File: C++/20. Valid Parentheses/Solution.js
var isValid = function(s) {
    const str = s.split('');
    let check = [];

    for (let i = 0; i < str.length; i++) {
        if (str[i] === '(' || str[i] === '[' || str[i] === '{') {
            check.push(str[i]);
        } else if (str[i] === ')' || str[i] === ']' || str[i] === '}') {
            let last = check.pop();

            if (
                (str[i] === ')' && last !== '(') ||
                (str[i] === ']' && last !== '[') ||
                (str[i] === '}' && last !== '{')
            ) {
                return false; // Mismatch found, return false immediately
            }
        } else {
            return false; // Non-bracket character found, return false immediately
        }
    }

    return check.length === 0; // Return true if all brackets are matched
};

File: C++/20. Valid Parentheses/Solution.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> arr;
        int n = s.size();
        for(int i=0;i<n;i++){
            if(s[i]=='(' || s[i]=='[' || s[i]=='{'){
                arr.push(s[i]);
            }else {
                if(arr.empty()){
                    return false;
                }
                char last = arr.top();
                if (
                (s[i] == ')' && last != '(') ||
                (s[i] == ']' && last != '[') ||
                (s[i] == '}' && last != '{')){
                    return false;
                }
                arr.pop();
            }
        }
        return arr.empty();
    }
};
